RIPS/rip-9997.md

```diff
@@ -0,0 +1,208 @@
+
+```
+BUNDLE_HEADER_SUBTYPE = x
+```
+
+Instead, we propose to introduce an explicit "counter" transaction subtype.
+
+The payload should be interpreted as:
+
+```
+0x04 || 0x01 || rlp([chainId, transactionCount])
+```
+
+Header transactions have a unique hash calculated as follows:
+
+```
+keccak256(AA_TX_TYPE || 0x01 || rlp(chainId, transactionCount, blockNumber, txIndex))
+```
+
+The `transactionCount` is required to be strictly larger than `1`.
+
+The `blockNumber` and `txIndex` parameters are added to the hash to achieve unique header transaction IDs.
+
+The header transactions are only used to help execution clients determine how many of the `AA_TX_TYPE` transactions
+belong to each individual bundle.
+
+The block is not valid if a header transaction is located anywhere except before a set of `AA_TX_TYPE` transactions.
+
+Header transactions do not affect blockchain state and do not cost any gas.
+
+### Non-atomic validation and execution transaction subtype
+
+```
+BUNDLE_TRANSACTION_SUBTYPE = x
+```
+
+These transactions are completely identical to regular RIP-7560 transactions with the exceptions of being
+composable into bundles.
+
+It is important for the wallets to explicitly opt into this feature by accepting the `BUNDLE_TRANSACTION_SUBTYPE`,
+as some Smart Contract Accounts
+may be created in a way that relies on atomicity between validation and execution phases.
+
+### All validation state changes apply before all execution ones
+
+Filling a block with AA transactions must not be a challenge for the block builder.
+However, if each transaction during its execution can alter any state that affects the validity of another transaction
+in the mempool, the block builder will be forced to revalidate all transactions in the mempool after each inclusion.
+
+We mitigate that by applying all changes in all the validation frames of a sequence of AA transactions first
+and all execution frames apply immediately after that.
+
+In theory, the validation frames can also invalidate each other, but we define ways to prevent that by applying
+certain rules for the mempool transactions in ERC-7562.
+
+A builder that chooses not to enforce the rules from ERC-7562 **must** take care to re-validate each transaction
+against the mid-block state at the position where it is being included into a block.
+Otherwise, the resulting block is likely to end up being invalid.
+
+### Block structure diagram
+
+Here is a visual representation of a block that contains multiple Account Abstraction Transactions.
+The validation parts of AA transactions are executed as separate transactions,
+but are not represented as separate transactions in the block data.
+
+![](../assets/rip-7560/block_overview.png)
+*The structure of a block containing multiple Native Account Abstraction Transactions*
+
+### Validation state change virtual transactions
+
+The validation frames of the AA_TX_TYPE transaction are represented as individual virtual transactions by the clients.
```

Comment:
This is the first time a "virtual transaction" is mentioned. Maybe a definition would be helpful here?